pkg/controllers/jobset_controller.go

```diff
@@ -286,6 +253,155 @@ func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, js *
 	return nil
 }
 
+func (r *JobSetReconciler) executeFailurePolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("jobset", klog.KObj(js))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// If no failure policy is defined, the default failure policy is to mark the JobSet
+	// as failed if any of its jobs have failed.
+	if js.Spec.FailurePolicy == nil {
+		return r.updateStatusCondition(ctx, js, metav1.Condition{
+			Type:               string(jobset.JobSetFailed),
+			Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+			Reason:             "FailedJobs",
+			Message:            "jobset failed due to one or more failed jobs",
+		})
+	}
+
+	// Handle different types of failure policy targets.
+	switch js.Spec.FailurePolicy.Operator {
+	case jobset.TerminationPolicyTargetAny:
+		// To reach this point a job must have failed, and TerminationPolicyTargetAny applies to any job
+		// in the JobSet, so we can skip directly to executing the restart policy.
+		return r.failureRestartPolicy(ctx, js, ownedJobs)
+	default:
+		return fmt.Errorf("invalid termination policy target %s", js.Spec.FailurePolicy.Operator)
+	}
+}
+
+func (r *JobSetReconciler) failureRestartPolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	switch js.Spec.FailurePolicy.RestartPolicy {
+	case jobset.RestartPolicyRecreateAll:
+		return r.restartPolicyRecreateAll(ctx, js, ownedJobs)
+	case jobset.RestartPolicyNone:
+		return r.updateStatusCondition(ctx, js, metav1.Condition{
+			Type:               string(jobset.JobSetFailed),
+			Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+			Reason:             "FailedJobs",
+			Message:            "jobset failed due to one or more job failures",
+		})
+	default:
+		return fmt.Errorf("invalid restart policy: %s", js.Spec.FailurePolicy.RestartPolicy)
+	}
+}
+
+func (r *JobSetReconciler) restartPolicyRecreateAll(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("jobset", klog.KObj(js))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// If JobSet has reached max number of restarts, mark it as failed and return.
+	if js.Status.Restarts == js.Spec.FailurePolicy.MaxRestarts {
+		return r.updateStatusCondition(ctx, js, metav1.Condition{
+			Type:               string(jobset.JobSetFailed),
+			Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+			Reason:             "ReachedMaxRestarts",
+			Message:            "jobset failed due to reaching max number of restarts",
+		})
+	}
+
+	// Increment JobSet restarts. This will trigger reconciliation and result in deletions
+	// of old jobs not part of the current jobSet lifecycle.
+	js.Status.Restarts += 1
+	return r.Status().Update(ctx, js)
```

Comment:
log info here that the jobset is attempting restart number x after successfully returning from the update call